File: examples/posix-thread/example05.cpp
// =================================================================
//
// File: example05.cpp
// Author: Pedro Perez
// Description: This file contains the code that implements the
//				bubble sort algorithm. The time this implementation takes
//				will be used as the basis to calculate the improvement
//				obtained with parallel technologies.
//
// Copyright (c) 2022 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <cstring>
#include <pthread.h>
#include "utils.h"

const int SIZE = 10000; //1e5
const int THREADS = 4;

typedef struct {
  int start, end; // [start, end)
  int *arr;
} Block;

using namespace std;

void swap(int *A, int i, int j) {
  int aux = A[i];
  A[i] = A[j];
  A[j] = aux;
}

int* oddEvenSort(int *A, int size) {
    int *B = new int[size];

    memcpy(B, A, sizeof(int) * size);
	for (int step = 0; step < size; step++) {
		if (step % 2 == 0) {
			for (int i = 0; i <= size - 2; i += 2) {
				if (B[i] > B[i + 1]) {
					swap(B, i, i + 1);
				}
			}
		} else {
			for (int i = 1; i <= size - 2; i += 2) {
				if (B[i] > B[i + 1]) {
					swap(B, i, i + 1);
				}
			}
		}
	}
    return B;
}

int main(int argc, char* argv[]) {
	int *a, *b;
	double ms;

	a = new int[SIZE];
	random_array(a, SIZE);
	display_array("before", a);

	cout << "Starting..." << endl;
	ms = 0;
	for (int i = 0; i < N; i++) {
		start_timer();

		b = oddEvenSort(a, SIZE);

		ms += stop_timer();

        if (i != N - 1) {
			delete [] b;
		}
	}
	display_array("after", b);
	cout << "avg time = " << setprecision(5) << (ms / N) << " ms" << endl;

	delete [] a;
	delete [] b;
	return 0;
}

File: examples/processes-python/example05.py
## =================================================================
##
## File: example05.py
## Author: Pedro Perez
## Description: This file contains the code that implements the
##				bubble sort algorithm using processes in Python.
##
## Copyright (c) 2022 by Tecnologico de Monterrey.
## All Rights Reserved. May be reproduced for any non-commercial
## purpose.
##
## =================================================================

import utils
import time
import math
import multiprocessing as mp
import threading

# CORES = mp.cpu_count() 
CORES = 8
SIZE = 10000 ##1e4

def swap(array, i, j):
    aux = array[i]
    array[i] = array[j]
    array[j] = aux

def bubbleSort(start, end, array):
    for i in range(end - 1, start - 1, -1):
        for j in range(start, i):
            if array[j] > array[j + 1]:
                swap(array, j, j + 1)

def mergeAndCopy(start, mid, end, array, temp):
    i = start
    j = mid
    k = start
    while i < mid and j < end:
        if array[i] < array[j]:
            temp[k] = array[i]
            i += 1
        else:
            temp[k] = array[j]
            j += 1
        k += 1

    while i < mid:
        temp[k] = array[i]
        i += 1
        k += 1

    while j < end:
        temp[k] = array[j]
        j += 1
        k += 1

    for i in range(start, end):
        array[i] = temp[i]
            
def partialSort(depth, start, end, array, temp):
    if depth == 0:
        bubbleSort(start, end, array)
    else:
        mid = start + ((end - start) // 2)
        left = mp.Process(target=partialSort, args=(depth - 1, start, mid, array, temp, ))  
        right = mp.Process(target=partialSort, args=(depth - 1, start, mid, array, temp, ))

        left.start()
        right.start()

        left.join()
        right.join()
        
        mergeAndCopy(start, mid, end, array, temp)

if __name__ == "__main__":
    array = mp.Array('i', range(SIZE))
    temp = mp.Array('i', range(SIZE))

    utils.randomArray(array)
    utils.displayArray("array", array)

    print("Starting...")
    startTime = endTime = ms = 0
    for j in range(utils.N):
        startTime = time.time() * 1000

        process = mp.Process(target=partialSort, args=(CORES, 0, SIZE, array, temp, ))  
        process.start()
        process.join()

        endTime = time.time() * 1000

        ms = ms + (endTime - startTime)

    utils.displayArray("array", array)
    print("avg time = ", (ms / utils.N), " ms")
